modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java

```diff
@@ -177,6 +179,31 @@ public final void collectMetrics(Request<?> request, Response<?> response) {
                 .map(List::size)
                 .orElse(0);
 
+            if (exceptionCount > 0) {
```

Comment:
I'm continuing the conversation on how the stats should look like here @ywangd @arteam . I believe an ideal approach would be to:

- `METRIC_EXCEPTIONS_TOTAL` counts the server-related errors (like 5xx)
- This PR's new `METRIC_CLIENT_EXCEPTIONS_TOTAL` to count the client-related errors (like 4xx)
- And to be able to somehow later filter out specific error codes (like 416) from the client-related errors

I'm not sure what's the best way to achieve that. I can think of a couple:

- Either a separate metric/LongCounter for each status error code.
- Or a "total" counter (as is now) but with added attribute(s) that should hopefully allows us to filter out specific status codes.

To avoid adding a gazillion counters for each status code, I have a preference for the latter approach. But unsure whether it can work in practice, i.e., when you go to Kibana later, that you can indeed easily e.g. filter out the 416s.

Thus, I'd suggest:

- Count all the client-related exceptions in `METRIC_CLIENT_EXCEPTIONS_TOTAL`.
- Add each client error code of `statusCodes` to the attributes, with their sum. E.g., if statusCodes is [429, 416, 429, 429], then the added attributes would be `key=429, value=3` and `key=416, value=1`.
- Spin up a custom cloud or serverless deployment out of this PR, and execute a workload that will provoke some of these. I believe a rally benchmark should be able to provoke some 416s.
- Go to Kibana and make two dashboard visualizations: (a) one that shows ALL client errors, and (b) one that shows/plots/sums only 416s. This will confirm we can do the filtering using the attributes.
- If we confirm the approach is working, we can do the same to add the attributes also to `METRIC_CLIENT_EXCEPTIONS_HISTOGRAM`, and I could also see adding server-related status codes attributes to `METRIC_EXCEPTIONS_TOTAL`/`METRIC_EXCEPTIONS_HISTOGRAM`.

What do you think?

Reply:
@kingherc I believe the current setup should work without additional counters.

* I'd like to keep this PR focused on `RequestedRangeNotSatisfiedExceptions`, but I'm happy to adjust the code to handle all client errors after we verify that counting 416 errors works. 

* Filtering by attributes works. The error code is added to metrics under the `labels` prefix and the metric has a name `es.repositories.exceptions.client.total.labels.error_code` with the value 416. We just adjust the KQL query add an additional clause.

* We already add the labels to the client error histogram. If this would work for 416 errors, I'm happy to adjust it to cover all 4xx error codes.

Reply:
@arteam I see, thanks for explaining. Makes sense to limit this PR to just RequestedRangeNotSatisfiedExceptions!

However, the current PR is a bit "mixed". For example some points of incoherence:

- `METRIC_CLIENT_EXCEPTIONS_TOTAL` only counts 416s.
- `METRIC_EXCEPTIONS_TOTAL` includes client exceptions.

To make it simpler and clearer, maybe we should simply rename `METRIC_CLIENT_EXCEPTIONS_TOTAL` to `METRIC_REQUESTED_RANGE_NOT_SATISFIED_EXCEPTIONS_TOTAL` and similarly for the histogram one. What do you think? And in the future, if we'd like to capture all kinds of client exceptions separately, we can issue a new ticket and start a new discussion.

Reply:
@kingherc Sounds good to me! That was the name of metrics in the first iteration of this pull request, I've reverted the commits to it, but kept adding the error code as a metric attribute, so we can adapt the metrics to all client errors in the future.

Reply:
I am not sure that I follow the decision here. IIUC, we agreed that different exceptions can be filtered/differentiated at reporting time (in Kibana). In that case, shouldn't we use a single metric, i.e. the existing `METRIC_EXCEPTIONS_TOTAL` for the exceptions and just add labels for the `error_code`. I don't quite understand why we decided to have a separate `METRIC_CLIENT_EXCEPTIONS_TOTAL`?

Reply:
@ywangd I guess we we wanted to have a clear separation between exceptions and client errors based on status codes.
The biggest issue is that we can multiple status codes, so you would need to aggregate them before adding as labels, and I'm not sure how that would work with histograms, for example.

I think the current implementation solves the issue in the Jira ticket and also verifies that attributes work. We can probably iterate on a generic solutions for client errors in a follow up issue?

Reply:
@ywangd , @arteam , I would also prefer to holistically handle the error codes so we can filter them separately, but judging from the long conversations we're having on how to make that happen (whether with labels or separate counters or some other way), I believe it may be another ticket/PR altogether. 

So I think indeed that this PR that introduces just a separate counter for the REQUESTED_RANGE_NOT_SATISFIED exceptions is sufficient to resolve ES-8636.

What do you think?

Reply:
Potentially we can create a new ticket for all error codes, and when it's done, we can remove the specific metrics about REQUESTED_RANGE_NOT_SATISFIED exceptions introduced by this PR. ?

Reply:
I totally agree! I think that is the best way forward. I can create a ticket to track all client errors with labels, but keep a dedicated `REQUESTED_RANGE_NOT_SATISFIED` metric for the time being which solves the issue in the ticket.

Reply:
I still think using labels/attributes for differentiating the different exception is better. That said, it is something that we can refine relatively easily later. So I am onboard with the current approach.